File: ChessMateEngine/Movement/Move.cpp
#include "Move.h"

namespace Movement {

//==============================================================================
Move::Move() :
    m_startRank(-1),
    m_startFile(-1),
    m_endRank(-1),
    m_endFile(-1),
    m_movingPiece(-1),
    m_promotionPiece(-1),
    m_IsCheck(false),
    m_IsCheckmate(false),
    m_IsStalemate(false),
    m_IsCapture(false),
    m_IsKingsideCastle(false),
    m_IsQueensideCastle(false),
    m_IsEnPassant(false)
{
}

//==============================================================================
Move::Move(
    const Game::square_type &sRank,
    const Game::square_type &sFile,
    const Game::square_type &eRank,
    const Game::square_type &eFile
) :
    m_startRank(sRank),
    m_startFile(sFile),
    m_endRank(eRank),
    m_endFile(eFile),
    m_movingPiece(-1),
    m_promotionPiece(-1),
    m_IsCheck(false),
    m_IsCheckmate(false),
    m_IsStalemate(false),
    m_IsCapture(false),
    m_IsKingsideCastle(false),
    m_IsQueensideCastle(false),
    m_IsEnPassant(false)
{
}

//==============================================================================
Move::Move(
    const Game::square_type &sRank,
    const Game::square_type &sFile,
    const Game::square_type &eRank,
    const Game::square_type &eFile,
    const Game::piece_type &promotionPiece
) :
    m_startRank(sRank),
    m_startFile(sFile),
    m_endRank(eRank),
    m_endFile(eFile),
    m_movingPiece(-1),
    m_promotionPiece(promotionPiece),
    m_IsCheck(false),
    m_IsCheckmate(false),
    m_IsStalemate(false),
    m_IsCapture(false),
    m_IsKingsideCastle(false),
    m_IsQueensideCastle(false),
    m_IsEnPassant(false)
{
}

//==============================================================================
Move::Move(const std::string &move, Game::color_type player) :
    m_promotionPiece(Game::PAWN),
    m_IsCheck(false),
    m_IsCheckmate(false),
    m_IsStalemate(false),
    m_IsCapture(false),
    m_IsKingsideCastle(false),
    m_IsQueensideCastle(false),
    m_IsEnPassant(false)
{
    // Parse for castle
    if (move.compare("O-O") == 0)
    {
        m_movingPiece = Game::KING;
        m_IsKingsideCastle = true;

        m_startFile = Game::FILE_E;
        m_endFile = Game::FILE_G;

        if (player == Game::WHITE)
        {
            m_startRank = Game::RANK_1;
            m_endRank = Game::RANK_1;
        }
        else
        {
            m_startRank = Game::RANK_8;
            m_endRank = Game::RANK_8;
        }

        return;
    }
    else if (move.compare("O-O-O") == 0)
    {
        m_movingPiece = Game::KING;
        m_IsQueensideCastle = true;

        m_startFile = Game::FILE_E;
        m_endFile = Game::FILE_C;

        if (player == Game::WHITE)
        {
            m_startRank = Game::RANK_1;
            m_endRank = Game::RANK_1;
        }
        else
        {
            m_startRank = Game::RANK_8;
            m_endRank = Game::RANK_8;
        }

        return;
    }

    unsigned int ix = 0;

    // Parse starting piece
    if (move[ix] == 'N')
    {
        m_movingPiece = Game::KNIGHT;
        ++ix;
    }
    else if (move[ix] == 'B')
    {
        m_movingPiece = Game::BISHOP;
        ++ix;
    }
    else if (move[ix] == 'R')
    {
        m_movingPiece = Game::ROOK;
        ++ix;
    }
    else if (move[ix] == 'Q')
    {
        m_movingPiece = Game::QUEEN;
        ++ix;
    }
    else if (move[ix] == 'K')
    {
        m_movingPiece = Game::KING;
        ++ix;
    }
    else
    {
        m_movingPiece = Game::PAWN;
    }

    // Parse starting position
    m_startFile = move[ix++] - 'a';
    m_startRank = move[ix++] - '1';

    // Parse for capture
    if (move[ix] == 'x')
    {
        ix++;
        m_IsCapture = true;
    }

    // Parse ending position
    m_endFile = move[ix++] - 'a';
    m_endRank = move[ix++] - '1';

    // Special moves
    while (ix < move.length())
    {
        if (move[ix] == '=')
        {
            ++ix;

            if (move[ix] == 'N')
            {
                m_promotionPiece = Game::KNIGHT;
            }
            else if (move[ix] == 'B')
            {
                m_promotionPiece = Game::BISHOP;
            }
            else if (move[ix] == 'R')
            {
                m_promotionPiece = Game::ROOK;
            }
            else if (move[ix] == 'Q')
            {
                m_promotionPiece = Game::QUEEN;
            }
            else
            {
                m_promotionPiece = Game::PAWN;
            }
        }
        else if (move[ix] == 'e')
        {
            if (move[++ix] == 'p')
            {
                m_IsEnPassant = true;
            }
        }
        else if (move[ix] == '+')
        {
            m_IsCheck = true;
        }
        else if (move[ix] == '#')
        {
            m_IsCheckmate = true;
        }

        ++ix;
    }
}

//==============================================================================
Game::square_type Move::GetStartFile() const
{
    return m_startFile;
}

//==============================================================================
Game::square_type Move::GetStartRank() const
{
    return m_startRank;
}

//==============================================================================
Game::square_type Move::GetEndFile() const
{
    return m_endFile;
}

//==============================================================================
Game::square_type Move::GetEndRank() const
{
    return m_endRank;
}

//==============================================================================
std::string Move::GetPGNString() const
{
    if (m_IsKingsideCastle)
    {
        return "O-O";
    }
    else if (m_IsQueensideCastle)
    {
        return "O-O-O";
    }

    std::string ret("");

    // Set the moving piece
    if (m_movingPiece == Game::KNIGHT)
    {
        ret += "N";
    }
    else if (m_movingPiece == Game::BISHOP)
    {
        ret += "B";
    }
    else if (m_movingPiece == Game::ROOK)
    {
        ret += "R";
    }
    else if (m_movingPiece == Game::QUEEN)
    {
        ret += "Q";
    }
    else if (m_movingPiece == Game::KING)
    {
        ret += "K";
    }

    // Append the start file and start rank
    ret += ('a' + m_startFile);
    ret += ('1' + m_startRank);

    // If move is capture, indicate such
    if (m_IsCapture)
    {
        ret += 'x';
    }

    // Append the end file and end rank
    ret += ('a' + m_endFile);
    ret += ('1' + m_endRank);

    // If move is a promotion, indicate the promotion piece
    if (m_promotionPiece == Game::KNIGHT)
    {
        ret += "=N";
    }
    else if (m_promotionPiece == Game::BISHOP)
    {
        ret += "=B";
    }
    else if (m_promotionPiece == Game::ROOK)
    {
        ret += "=R";
    }
    else if (m_promotionPiece == Game::QUEEN)
    {
        ret += "=Q";
    }

    // If move is an enpassant, indicate such
    if (m_IsEnPassant)
    {
        ret += "ep";
    }

    // If move is check or checkmate, indicate such
    if (m_IsCheckmate)
    {
        ret += '#';
    }
    else if (m_IsCheck)
    {
        ret += '+';
    }

    return ret;
}

//==============================================================================
Game::piece_type Move::GetMovingPiece() const
{
    return m_movingPiece;
}

//==============================================================================
void Move::SetMovingPiece(Game::piece_type movingPiece)
{
    m_movingPiece = movingPiece;
}

//==============================================================================
Game::piece_type Move::GetPromotionPiece() const
{
    return m_promotionPiece;
}

//==============================================================================
void Move::SetPromotionPiece(Game::piece_type promotionPiece)
{
    m_promotionPiece = promotionPiece;
}

//==============================================================================
bool Move::IsCapture() const
{
    return m_IsCapture;
}

//==============================================================================
void Move::SetCapture()
{
    m_IsCapture = true;
}

//==============================================================================
bool Move::IsCheck() const
{
    return m_IsCheck;
}

//==============================================================================
void Move::SetCheck()
{
    m_IsCheck = true;
}

//==============================================================================
bool Move::IsCheckmate() const
{
    return m_IsCheckmate;
}

//==============================================================================
void Move::SetCheckmate()
{
    m_IsCheckmate = true;
}

//==============================================================================
bool Move::IsEnPassant() const
{
    return m_IsEnPassant;
}

//==============================================================================
void Move::SetEnPassant()
{
    m_IsEnPassant = true;
}

//==============================================================================
bool Move::IsKingsideCastle() const
{
    return m_IsKingsideCastle;
}

//==============================================================================
void Move::SetKingsideCastle()
{
    m_IsKingsideCastle = true;
}

//==============================================================================
bool Move::IsQueensideCastle() const
{
    return m_IsQueensideCastle;
}

//==============================================================================
void Move::SetQueensideCastle()
{
    m_IsQueensideCastle = true;
}

//==============================================================================
bool Move::operator == (const Move &move) const
{
    return
    (
        (m_startRank == move.m_startRank) &&
        (m_startFile == move.m_startFile) &&
        (m_endRank == move.m_endRank) &&
        (m_endFile == move.m_endFile)
    );
}

//==============================================================================
std::ostream &operator << (std::ostream &stream, const Move &move)
{
    stream << move.GetPGNString();
    return stream;
}

}

File: ChessMateGUI/src/main/java/com/flynn/chessmate/movement/Move.java
package com.flynn.chessmate.movement;

import com.flynn.chessmate.game.Piece;
import com.flynn.chessmate.util.Constants;

import java.awt.Color;

/**
 * Data structure to store information about a move.
 *
 * @author Timothy Flynn (timothy.flynn@outlook.com)
 * @version March 3, 2013
 */
public class Move
{
    private int m_startRank;
    private int m_startFile;
    private int m_endRank;
    private int m_endFile;

    private Piece m_movingPiece;

    private boolean m_isCheck;
    private boolean m_isCheckmate;
    private boolean m_isStalemate;
    private boolean m_isCapture;
    private boolean m_isKingsideCastle;
    private boolean m_isQueensideCastle;
    private boolean m_isEnPassant;
    private Piece m_pawnPromotionPiece;
    private int m_ambiguityType;

    /**
     * Constructor.
     *
     * @param rank The rank of the start position.
     * @param file The file of the start position.
     * @param i The rank of the end position.
     * @param j The file of the end position.
     */
    public Move(int rank, int file, int i, int j)
    {
        this.m_startRank = rank;
        this.m_startFile = file;
        this.m_endRank = i;
        this.m_endFile = j;

        m_isCheck = false;
        m_isCheckmate = false;
        m_isStalemate = false;
        m_isCapture = false;
        m_isKingsideCastle = false;
        m_isQueensideCastle = false;
        m_isEnPassant = false;
        m_pawnPromotionPiece = null;
        m_ambiguityType = 0;
    }

    /**
     * Constructor to parse a move in PGN format.
     *
     * @param move The PGN formatted string.
     * @param player The color of the player who generated the move.
     */
    public Move(String move, Color player)
    {
        // Parse for castle
        if (move.equals("O-O"))
        {
            m_isKingsideCastle = true;

            m_startFile = Constants.FILE_E;
            m_endFile = Constants.FILE_G;

            if (player == Color.WHITE)
            {
                m_movingPiece = Piece.WHITE_KING;
                m_startRank = Constants.RANK_1;
                m_endRank = Constants.RANK_1;
            }
            else
            {
                m_movingPiece = Piece.BLACK_KING;
                m_startRank = Constants.RANK_8;
                m_endRank = Constants.RANK_8;
            }

            return;
        }
        else if (move.equals("O-O-O"))
        {
            m_isQueensideCastle = true;

            m_startFile = Constants.FILE_E;
            m_endFile = Constants.FILE_C;

            if (player == Color.WHITE)
            {
                m_movingPiece = Piece.WHITE_KING;
                m_startRank = Constants.RANK_1;
                m_endRank = Constants.RANK_1;
            }
            else
            {
                m_movingPiece = Piece.BLACK_KING;
                m_startRank = Constants.RANK_8;
                m_endRank = Constants.RANK_8;
            }

            return;
        }

        int ix = 0;

        // Parse moving piece
        if (move.startsWith("N"))
        {
            ix++;
            m_movingPiece = (player == Color.WHITE ? Piece.WHITE_KNIGHT : Piece.BLACK_KNIGHT);
        }
        else if (move.startsWith("B"))
        {
            ix++;
            m_movingPiece = (player == Color.WHITE ? Piece.WHITE_BISHOP : Piece.BLACK_BISHOP);
        }
        else if (move.startsWith("R"))
        {
            ix++;
            m_movingPiece = (player == Color.WHITE ? Piece.WHITE_ROOK : Piece.BLACK_ROOK);
        }
        else if (move.startsWith("Q"))
        {
            ix++;
            m_movingPiece = (player == Color.WHITE ? Piece.WHITE_QUEEN : Piece.BLACK_QUEEN);
        }
        else if (move.startsWith("K"))
        {
            ix++;
            m_movingPiece = (player == Color.WHITE ? Piece.WHITE_KING : Piece.BLACK_KING);
        }
        else
        {
            m_movingPiece = (player == Color.WHITE ? Piece.WHITE_PAWN : Piece.BLACK_PAWN);
        }

        // Parse starting position
        m_startFile = move.charAt(ix++) - 'a';
        m_startRank = move.charAt(ix++) - '1';

        // Parse for capture
        if (move.contains("x"))
        {
            ix++;
            m_isCapture = true;
        }

        // Parse ending position
        m_endFile = move.charAt(ix++) - 'a';
        m_endRank = move.charAt(ix++) - '1';

        // Parse pawn promotion
        if (move.contains("=N"))
        {
            m_pawnPromotionPiece =
                (player == Color.WHITE ? Piece.WHITE_KNIGHT : Piece.BLACK_KNIGHT);
        }
        else if (move.contains("=B"))
        {
            m_pawnPromotionPiece =
                (player == Color.WHITE ? Piece.WHITE_BISHOP : Piece.BLACK_BISHOP);
        }
        else if (move.contains("=R"))
        {
            m_pawnPromotionPiece = (player == Color.WHITE ? Piece.WHITE_ROOK : Piece.BLACK_ROOK);
        }
        else if (move.contains("=Q"))
        {
            m_pawnPromotionPiece = (player == Color.WHITE ? Piece.WHITE_QUEEN : Piece.BLACK_QUEEN);
        }

        // Parse for enpassant
        if (move.contains("ep"))
        {
            m_isEnPassant = true;
        }

        // Parse for check / checkmate
        if (move.contains("+"))
        {
            m_isCheck = true;
        }
        else if (move.contains("#"))
        {
            m_isCheckmate = true;
        }
    }

    /**
     * @return A string representation of the move.
     */
    public String getPgnString(boolean preserveAmbiguity)
    {
        if (m_isKingsideCastle)
        {
            return "O-O";
        }
        else if (m_isQueensideCastle)
        {
            return "O-O-O";
        }

        StringBuilder move = new StringBuilder();

        // Set the moving piece
        switch (m_movingPiece)
        {
            case WHITE_KNIGHT:
            case BLACK_KNIGHT:
                move.append("N");
                break;
            case WHITE_BISHOP:
            case BLACK_BISHOP:
                move.append("B");
                break;
            case WHITE_ROOK:
            case BLACK_ROOK:
                move.append("R");
                break;
            case WHITE_QUEEN:
            case BLACK_QUEEN:
                move.append("Q");
                break;
            case WHITE_KING:
            case BLACK_KING:
                move.append("K");
                break;
            default:
                break;
        }

        // Append the start file and start rank
        if (!preserveAmbiguity || (m_ambiguityType & 1) > 0)
        {
            move.append((char)('a' + m_startFile));
        }
        if (!preserveAmbiguity || (m_ambiguityType & 2) > 0)
        {
            move.append(String.valueOf(m_startRank + 1));
        }

        // If move is capture, indicate such
        if (m_isCapture)
        {
            move.append("x");
        }

        // Append the end file and end rank
        move.append((char)('a' + m_endFile));
        move.append(String.valueOf(m_endRank + 1));

        // If move is a promotion, indicate the promotion piece
        if (m_pawnPromotionPiece != null)
        {
            if (m_pawnPromotionPiece != Piece.WHITE_PAWN &&
                m_pawnPromotionPiece != Piece.BLACK_PAWN)
            {
                switch (m_pawnPromotionPiece)
                {
                    case WHITE_KNIGHT:
                    case BLACK_KNIGHT:
                        move.append("=N");
                        break;
                    case WHITE_BISHOP:
                    case BLACK_BISHOP:
                        move.append("=B");
                        break;
                    case WHITE_ROOK:
                    case BLACK_ROOK:
                        move.append("=R");
                        break;
                    case WHITE_QUEEN:
                    case BLACK_QUEEN:
                        move.append("=Q");
                        break;
                    default:
                        break;
                }
            }
        }

        // If move is an enpassant, indicate such
        if (m_isEnPassant)
        {
            move.append("ep");
        }

        // If move is check or checkmate, indicate such
        if (m_isCheckmate)
        {
            move.append("#");
        }
        else if (m_isCheck)
        {
            move.append("+");
        }

        return move.toString();
    }

    /**
     * @return The rank of the start position.
     */
    public int getStartRank()
    {
        return m_startRank;
    }

    /**
     * @return The file of the start position.
     */
    public int getStartFile()
    {
        return m_startFile;
    }

    /**
     * @return The rank of the end position.
     */
    public int getEndRank()
    {
        return m_endRank;
    }

    /**
     * @return The file of the end position.
     */
    public int getEndFile()
    {
        return m_endFile;
    }

    /**
     * Set the moving piece for this move.
     *
     * @param movingPiece The moving piece.
     */
    public void setMovingPiece(Piece movingPiece)
    {
        this.m_movingPiece = movingPiece;
    }

    /**
     * @return The piece being moved.
     */
    public Piece getMovingPiece()
    {
        return m_movingPiece;
    }

    /**
     * If this move promotes a pawn, set the piece being promoted to.
     *
     * @param pawnPromotionPiece The piece to promote to.
     */
    public void setPawnPromotionPiece(Piece pawnPromotionPiece)
    {
        this.m_pawnPromotionPiece = pawnPromotionPiece;
    }

    /**
     * @return The piece to promote to, or null.
     */
    public Piece getPawnPromotionPiece()
    {
        return m_pawnPromotionPiece;
    }

    /**
     * Set that this move is a capturing move.
     */
    public void setCapture()
    {
        this.m_isCapture = true;
    }

    /**
     * @return True if the move is a capturing move, false otherwise.
     */
    public boolean isCapture()
    {
        return m_isCapture;
    }

    /**
     * Set that this move puts the opponent in check.
     */
    public void setCheck()
    {
        this.m_isCheck = true;
    }

    /**
     * @return True if the opponent is in check, false otherwise
     */
    public boolean isCheck()
    {
        return m_isCheck;
    }

    /**
     * Set that this move puts the opponent in checkmate.
     */
    public void setCheckmate()
    {
        this.m_isCheckmate = true;
    }

    /**
     * @return True if the opponent is in checkmate, false otherwise
     */
    public boolean isCheckmate()
    {
        return m_isCheckmate;
    }

    /**
     * Set that this move puts the board in stalemate.
     */
    public void setStalemate()
    {
        this.m_isStalemate = true;
    }

    /**
     * @return True if the board is in stalemate, false otherwise
     */
    public boolean isStalemate()
    {
        return m_isStalemate;
    }

    /**
     * Set that this move is an en passant.
     *
     * @param isEnPassant True if the move is an en passant, false otherwise.
     */
    public void setEnPassant()
    {
        this.m_isEnPassant = true;
    }

    /**
     * @return True if this move is an en passant.
     */
    public boolean isEnPassant()
    {
        return m_isEnPassant;
    }

    /**
     * Set that this move is a kingside castle.
     */
    public void setKingsideCastle()
    {
        this.m_isKingsideCastle = true;
    }

    /**
     * @return True if this move is a kingside castle.
     */
    public boolean isKingsideCastle()
    {
        return m_isKingsideCastle;
    }

    /**
     * Set that this move is a queenside castle.
     */
    public void setQueensideCastle()
    {
        this.m_isQueensideCastle = true;
    }

    /**
     * @return True if this move is a queenside castle.
     */
    public boolean isQueensideCastle()
    {
        return m_isQueensideCastle;
    }

    /**
     * Set the ambiguity type for the move. Use the following values:
     * <li>0 = No ambiguity.
     * <li>1 = Resolve ambiguity by indicating file.
     * <li>2 = Resolve ambiguity by indicating rank.
     * <li>3 = Resolve ambiguity by indicating file and rank.
     *
     * @param ambiguityType Integer representing the type of ambiguity.
     */
    public void setAmbiguityType(int ambiguityType)
    {
        m_ambiguityType = ambiguityType;
    }
}
